### DISCUSSION TITLE
Relinking between two machines

### DISCUSSION THREAD

**Discussion by jhong75** [latenitefilms/BRAWToolbox#98] [Support] [answered]
I work remotely bringing a raided drive with me from work to home and vice versa. Peculiar behavior I've noticed is that I need to relink the BRAW material in an edited sequence between machines, even though the project/libary and media are residing on the same raided drive that I bring with me.

**Comment by jhong75** [comment]
Forgot to mention I'm running the current versions BRAW Toolbox 1.0.6

**Reply by tangierc** [reply]
Similar issue. I just upgraded to a new computer, but all of my media is on the same drives. I downloaded a fresh copy of BRAW Toolbox, but all of my clips are offline.
![Screenshot 2023-03-22 at 8 05 29 PM](https://user-images.githubusercontent.com/9577806/227097811-25f9e838-a7b1-4632-917a-9b18d61e5e3d.png)

**Comment by latenitefilms** [comment, marked as answer]
Assuming you're using the same RAID on two machines, try granting sandbox access on both machines, as explained here:

https://brawtoolbox.io/collaboration/

**Reply by jhong75** [reply]
I'll give this a try the next time I run into this situation! Thanks.

**Comment by latenitefilms** [comment]
For anyone who stumbles across this Discussion, v1.1.3 has an improved **Relink BRAW Clips within an LIBRARY / EVENT / PROJECT** Toolbox. You can learn more [here](https://brawtoolbox.io/toolbox/#relink-braw-clips-within-an-library--event--project).

**Reply by tangierc** [reply]
Hey Chris. I think there's some things that aren't clear in the instructions about the relinking process. I am in a similar situation as I am trying to relink my media that's been moved to a different drive.

Is there a way to relink all clips within the same directory at once or is it a clip by clip process? I dragged in a project timeline which has clips in different directories. When I paste the new directory path and hit process and replace, it updates once for one directory of clips, but other directories do not do the same thing. They do not update the path in the toolbox window. 

It's also not clear if you only have to update the path once for clips in a directory and by hitting Process and Find, does this automatically find all of the clips in the project for you and update the path. It seems so, but after the first time, I couldn't move on to subsequent clips and repeat the process. As a result of this, it seems that let's say if I have 20 directories of BRAW clips, that I'd have to process an new XML of my timeline 20 times before I get the timeline relinked properly. Though in my case I can't get the first XML processed without errors.

Since there's no direct link with a Finder item and the the BRAW Toolbox clips, does this process have to be done for every project in a an event / library since these are not global updates?

For now I am stuck relinking clips one by one again for one timeline.  The instructions include a lot of what to do, but there's a little bit of information missing about how this is supposed to work.

**Reply by tangierc** [reply]
Well you keep noodling around and you find more out. I think the problem is that I didn't realize that I had to manually make adjustments to the Find field and Replace field. I thought that by clicking or even double-clicking the clip with a red dot in the toolbox window, that this would load the shown path in the Find field automatically.

I am still not sure if I have to do this process for each timeline separately. Ideally that would not be the case. 

I'll keep playing with this some more.

**Comment by tangierc** [comment]
For those following I'll report that I have successfully relinked. To answer my own questions that someone else may wonder:

1. Yes, you have to manually change the Find field and Replace field and all instances of the contents of the Find field will be replaced - just like a text document
2. It's a good idea to hit Refresh and Request Permission after each successful Process Find and Replace, this way the red dots will be updated to green as you work through your relinking process.
3. Unfortunately only clips for one timeline updated so one would have to repeat this process for each timeline to relink...at least that's what happened in my case.

**Reply by latenitefilms** [reply]
When the Toolbox first opens, it displays a list of all the BRAW file paths. If any of the clips are red or orange, you'll need to manually correct the path to make them relink. For example, if you originally had your files at `/Volumes/Drive A/clip1.braw` but moved them to `/Volumes/Drive B/clip1.braw` you could do a **Find** for `Drive A` and replace it with `Drive B`. After updating the paths using the Find & Replace tool, you should then press **Refresh & Request Permissions** to make sure that everything re-links.

The Find & Replace Tool applies to ALL the paths in the list.

**Comment by tangierc** [comment]
Yep. It took me a minute to figure out how this is supposed to work, but I got it for one timeline. Problem is I have to do this for every timeline and every event. For a large feature film this is a huge task. It would mean repeating (in many cases) almost the same process of relinking timelines multiple times if one wants to reconnect all the media for the project timelines rather than do it once and it updates everywhere. My next thought was that the only way I can possibly get all of the BRAW clips relinked is to do it at the library level because then the toolbox would find/replace all of the same paths for the entire library. However when I tried this strategy my library won’t validate in the toolbox.  So right now I have one timeline mostly relinked successfully except for a few clips that aren’t BRAW clips, and a library with 20 events and many timeline not linked.  It seems my options are A) repeat the relink process for every timeline and event in a library or B) figure out how to get the library to validate and do it once so the clips are relinked throughout the entire library.

One work around would be to have all BRAW clips in one directory and that would reduce the amount of work, but not take care of the validation issue though. One would have to have foresight in the way the relinking toolbox works before getting started with the project though.

**Reply by latenitefilms** [reply]
You should be able to drag in **Libraries**, **Events** and **Projects** - so in theory, if it was working properly, you should just be able to drag in your entire library and it should work. Unfortunately Final Cut Pro itself doesn't actually do any validation when it exports a FCPXML, so it's entirely possible it could export an invalid FCPXML.

There's a macOS Terminal Command to validate a FCPXML file - you can learn more [here](https://fcp.cafe/developers/fcpxml/#dtd-validation).

What you could try do is try and work out WHY the FCPXML is failing, and then try to fix it in the library.

If you've just moved things from one drive to another, but kept the same folder structure, then, in theory, all you need to do is a find and replace for the volume name. For example, if you originally had your files at `/Volumes/Drive A/clip1.braw` but moved them to `/Volumes/Drive B/clip1.braw` you could do a **Find** for `Drive A` and replace it with `Drive B` - assuming that you don't have `Drive A` and `Drive B` also used in the file or folder names.

On this feature film you're using Synchronised Clips, so your Timeline BRAW Clips are independent from your Event Clips (i.e. if you rename a clip on your timeline, it won't affect the copy in the browser, etc). Therefore, to relink all your projects, you could just drag them all into the one event, then relink that event.

**Comment by tangierc** [comment]
I think I’ll give that last part a try.  I’ve tried cleaning XMLs in recent weeks to deal with any problem files, but there was always something not validating or giving an error. Apple did reach out to me recently for a related issue, but due to confidentiality instructions I can’t talk about it.

Tangier

**Reply by latenitefilms** [reply]
Looking at the FCPXML you sent me, this line seems to be the problem:

<img width="547" alt="image" src="https://github.com/latenitefilms/BRAWToolbox/assets/22286696/13c195ea-2267-4f06-a246-3eed6645d605">

Searching the FCPXML for that same reference ID brings me to this clip:

<img width="1190" alt="image" src="https://github.com/latenitefilms/BRAWToolbox/assets/22286696/e690f3ab-4a52-48b7-94f7-093c32e20114">

I'm assuming **A002_A001_02225Z_001** isn't a BRAW Toolbox clip?

**Reply by latenitefilms** [reply]
FWIW - It took Final Cut Pro a looooooong time to import, but when I did remove the **A002_A001_02225Z_001** instances in the FCPXML, it did actually import, with a bunch of **tcFormat** warnings.

<img width="432" alt="Screen Shot 2023-06-28 at 8 55 10 am" src="https://github.com/latenitefilms/BRAWToolbox/assets/22286696/718c4279-f592-4f94-92f8-06ef8d5823d8">

<img width="432" alt="Screen Shot 2023-06-28 at 8 55 36 am" src="https://github.com/latenitefilms/BRAWToolbox/assets/22286696/34fdb7d7-6bd2-4bcb-a04f-7c1ddabfabe1">

<img width="848" alt="image" src="https://github.com/latenitefilms/BRAWToolbox/assets/22286696/54164af9-afc7-4aca-b557-c7db650d9eba">

**Comment by tangierc** [comment]
It’s a Red clip. not in my timeline either. It’s in a mutlicam clip and the clip can’t be found or reconstructed. It’s just black. I’ll have to drop it back in there from the R3D file.

Strange, it’s RED file (made up of 32 R3D files) that FCP isn’t able to play. It’ll import, but not play video from it. It’s just black.  In REDCINEX-Pro it shows up just fine as does my other RED footage.

I’ll keep investigating.  

Thanks Chris.

**Reply by latenitefilms** [reply]
I think this is your problem clip which is causing FCPX to output invalid FCPXML. Once you remove that, it may fix things.

**Comment by tangierc** [comment]
Unfortunately every time I try dragging the event with with all of my project timelines, I see an exporting XML progress window, but ultimately it keeps failing, I see the dragging effect reverse to the library (somewhat ghosted), and BRAW Toolbox keeps saying Reopen Extension with an exclamation mark in an icon in the middle. Then the validation error pops up. That one problem clip is not in the event.

Tangier

**Reply by latenitefilms** [reply]
Weird. Final Cut Pro must be sending something that BRAW Toolbox isn't expecting. Can you share your crash logs and log files?

**Comment by tangierc** [comment]
Yeah, I am exporting the event XML and attempting to reimport it to see if something get’s flagged as a problem. FCP keeps hanging though Maybe Resolve will identify a problem with the XML.

**Comment by tangierc** [comment]
Vide App diagnostics sent.

**Reply by latenitefilms** [reply]
I can't see any Final Cut Pro crashes in that diagnostics file. There are two "hangs" - but if this is when you're dragging and dropping a Final Cut Pro library, you might just have to wait longer. I'm not sure how long it took to import your FCPXML, as I walked away from the computer, but it definitely took longer than 15mins - it could have been hours.

**Comment by tangierc** [comment]
Thanks. I got a lot of errors attempting to import as well. I copy pasted them yesterday to address them one by one. I’ve addressed them before and they were related to audio files and maybe one or two shutter stock files.  Still unclear why these are presenting issues.

These look like the same ones I encountered yesterday:

XML problem clips
Last Relink.fcpxml	Invalid edit with no respective media. (/fcpxml[1]/project[1]/sequence[1]/spine[1]/asset-clip[17])
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
Last Relink.fcpxml	Encountered an unexpected value. (tcFormat="DF": ***@***.***)
